Ray troubleshooting after Previewer number 2 (#186)
golemfactory/golem-docs: src/pages/docs/creators/ray/troubleshooting.md

type: troubleshooting
---

# Ray on Golem troubleshooting

{% troubleshooting %}

## What is going on with my cluster - status and dashboard 
 

{% problem /%}

Often you will find yourself wondering "How is my cluster doing?"

{% solution %}

Ray offers two tools to inspect the cluster state: the status and the dashboard.

Ray status shows information about nodes constituting the cluster, the total resources of the cluster, and their current usage.

The `ray status` command needs to be executed on the head node:
```bash
ray exec golem-cluster.yaml 'ray status'
```
```
Ray On Golem webserver
  Not starting webserver, as it's already running
Fetched IP: 192.168.0.3
======== Autoscaler status: 2023-11-28 08:09:48.391906 ========
Node status
---------------------------------------------------------------
Healthy:
 1 ray.head.default
 1 ray.worker.default
Pending:
 ray.worker.default, 5 launching
Recent failures:
 (no failures)

Resources
---------------------------------------------------------------
Usage:
 2.0/2.0 CPU
 0B/5.34GiB memory
 0B/2.61GiB object_store_memory

Demands:
 {'CPU': 1.0}: 385+ pending tasks/actors
Shared connection to 192.168.0.3 closed.

```

> The Ray dashboard visualizes the nodes and what happens to them. 

Reviewer: ```suggestion
Ray dashboard visualizes the nodes and various other aspects of a running cluster, like specific jobs and function calls that get executed by each node.
```

Run `ray dashboard golem-cluster.yaml` to forward your cluster's dashboard to [http://localhost:8265](http://localhost:8265).

> Keep it on to keep the tunnel active.

Reviewer: ```suggestion
Keep it running in the background so that the tunnel stays active as long as you need it.
```

(I mostly wanted to prevent the repetition of `keep`)

> For an in-debt description of the dashboard's amazing secrets check out the [Ray docs](https://docs.ray.io/en/latest/ray-observability/getting-started.html).

Reviewer: ```suggestion
For a more thorough description of the dashboard's amazing secrets, check out the [Ray docs](https://docs.ray.io/en/latest/ray-observability/getting-started.html).
```

in-depth?

Where possible, we try to display a meaningful error message - e.g. when `ray up` fails to start up
the Ray on Golem's webserver, we display a few last lines of the logs.

> If that's not enough to suggest a way to fix the issue investigating logs may prove more useful.

Reviewer: ```suggestion
If that's not enough to suggest a way to fix the issue, investigating the logs may provide the required details.
```

{% troubleshooting %}

## Starting over with a clean slate 

{% problem /%}

It may happen that something goes wrong and you wish to start over with a clean slate. 

{% solution %}

The first thing to do is `ray down` - it should be enough to clear the situation, but sadly it isn't always the case. 

> Let's first check if there are any orphaned components:

Reviewer: :+1:

  71258 ?        Sl     2:35 yagna
```

The above shows the `ray-on-golem` webserver and the `yagna` daemon are running.

> Note that Ray on Golem leaves the `yagna` daemon running on purpose - it stays connected to the Golem network maintaining current information about the providers so that when you start up your cluster again the nodes are found quicker. 

Reviewer: ```suggestion
Note that Ray on Golem leaves the `yagna` daemon running on purpose - it stays connected to the Golem network maintaining current information about the providers so that when you start up your cluster again the nodes are found more quickly. 
```

> With that in mind, we recommend killing only `ray-on-golem` and leaving `yagna` running. 

Reviewer: I'd put these lines after "The surest way", since only then do we mention killing those services...

Author: Left it where it was, but elaborated more (and removed killing ;)
```
The above shows the `ray-on-golem` webserver and the `yagna` daemon are running.

Note that Ray on Golem leaves the `yagna` daemon running on purpose - it stays connected to the Golem network maintaining current information about the providers so that when you start up your cluster again the nodes are found more quickly.

With that in mind, when starting over we recommend stopping ray-on-golem and leaving `yagna` running (but to get a truly clean slate you might want to stop `yagna` too).

The surest way to stop these services is to kill them (using the PID numbers as shown in the first column):
```

{% /troubleshooting %}

{% troubleshooting %}

## Head node not found

{% problem /%}

Sometimes, you might get `RuntimeError: Head node of cluster (golem-cluster) not found!` error message when attempting to run a ray cli command.

{% solution %}

This means, that you are trying to run a command that requires a cluster to be up (like `ray submit`), but the cluster is not alive.

> The simplest reason might be that you didn't successfully run `ray up`.

Reviewer: ```suggestion
The simplest reason might be that you hadn't successfully run `ray up`.
```

Checkout [the logs](/docs/creators/ray/troubleshooting#ray-on-golem-s-log-files) to see what happened.

> Another explanation might be that the cluster was running, but it stopped - please share the logs with us to help us prevent this from happening in the future.

Reviewer: ```suggestion
Another explanation might be that the cluster had been launched, but it stopped or crashed - please share the logs with us to help us prevent this from happening in the future.
```
